### PR TITLE
8328877: [JNI] The JNI Specification needs to address the limitations of integer UTF-8 String lengths

### PR DESCRIPTION
**Pull request by dholmes-ora**
This is the implementation of a new method added to the JNI specification.

From the CSR request:

The `GetStringUTFLength` function returns the length as a `jint` (`jsize`) value and so is limited to returning at most `Integer.MAX_VALUE`. But a Java string can itself consist of `Integer.MAX_VALUE` characters, each of which may require more than one byte to represent them in modified UTF-8 format.** It follows then that this function cannot return the correct answer for all String values and yet the specification makes no mention of this, nor of any possible error to report if this situation is encountered.

\*\*The modified UTF-8 format used by the VM can require up to six bytes to represent one unicode character, but six byte characters are stored as UTF16 surrogate pairs. Hence the most bytes per character is 3, and so the maximum length is 3*`Integer.MAX_VALUE`. With compact strings this reduces to 2*`Integer.MAX_VALUE`.

Solution

Deprecate the existing JNI `GetStringUTFLength` method noting that it may return a truncated length, and add a new method, JNI `GetStringUTFLengthAsLong` that returns the string length as a `jlong` value.

---

We also add a truncation warning to `GetStringUTFLength` under -Xcheck:jni

There are some incidental whitespace changes in `src/hotspot/os/posix/dtrace/hotspot_jni.d` along with the new method entries.

Testing:
 - new test added
 - tiers 1-3 sanity

Thanks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8338709](https://bugs.openjdk.org/browse/JDK-8338709) to be approved

### Issues
 * [JDK-8328877](https://bugs.openjdk.org/browse/JDK-8328877): [JNI] The JNI Specification needs to address the limitations of integer UTF-8 String lengths (**Enhancement** - P4)
 * [JDK-8338709](https://bugs.openjdk.org/browse/JDK-8338709): [JNI] The JNI Specification needs to address the limitations of integer UTF-8 String lengths (**CSR**)

### Reviewers
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**)
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20784/head:pull/20784` \
`$ git checkout pull/20784`

Update a local copy of the PR: \
`$ git checkout pull/20784` \
`$ git pull https://git.openjdk.org/jdk.git pull/20784/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20784`

View PR using the GUI difftool: \
`$ git pr show -t 20784`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20784.diff">https://git.openjdk.org/jdk/pull/20784.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20784#issuecomment-2319694538)

### CODE REVIEW

File: test/hotspot/jtreg/runtime/jni/checked/TestLargeUTF8Length.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/* @test
+ * @bug 8328877
+ * @summary Test warning for GetStringUTFLength and functionality of GetStringUTFLengthAsLong
+ * @library /test/lib
```
Comment (plummercj) [inline test/hotspot/jtreg/runtime/jni/checked/TestLargeUTF8Length.java:27]: Shouldn't this test have:

`@requires vm.bits == 64
`

Comment (dholmes-ora) [reply test/hotspot/jtreg/runtime/jni/checked/TestLargeUTF8Length.java:27]: Thanks for taking a look @plummercj . Yep I suppose it should.

**Review by plummercj — APPROVED** [review]
Overall it looks good to me, although I don't have experience adding a new JNI API (the dtrace probes were new to me), but it seems you are following what is already in place for other functions, and the testing looks good.

**Comment by dholmes-ora** [comment]
Thanks for the review @plummercj !

**Review by AlanBateman — APPROVED** [review]
Deprecating the existing function and introducing the new function looks okay.

The test is really 3 tests in one (GetStringUTFLength returning a truncated size, GetStringUTFLength with -Xcheck:jni prints a warning, and GetStringUTFLengthAsLong returns the long size). Personally I would have done this as 3 test cases rather in one launch with -Xcheck:jni but that's your choice.

**Comment by dholmes-ora** [comment]
Thanks for the review @AlanBateman . I like the test the way it is.

**Comment by dholmes-ora** [comment]
@plummercj  and @AlanBateman  could you please re-review as I was missing the main parts of the JNI version update! Thanks

**Review by AlanBateman — APPROVED** [review]

**Comment by dholmes-ora** [comment]
Thanks @AlanBateman !

**Review by plummercj — APPROVED** [review]

**Comment by dholmes-ora** [comment]
Thanks @plummercj !

**Pull request closed without merging**